crates/vm/src/system/memory/controller/mod.rs

```diff
@@ -463,46 +440,151 @@ impl<F: PrimeField32> MemoryController<F> {
     }
 
     /// Returns the final memory state if persistent.
+    #[allow(clippy::assertions_on_constants)]
     pub fn finalize<H>(&mut self, hasher: Option<&mut H>)
     where
         H: HasherChip<CHUNK, F> + Sync + for<'a> SerialReceiver<&'a [F]>,
     {
-        if self.final_state.is_some() {
-            return;
-        }
-        todo!();
-
+        let touched_blocks = self.memory.touched_blocks().collect::<Vec<_>>();
+        // First, let's split everything into blocks of ALIGN
+        let aligned_final_memory = {
+            let mut final_memory = TimestampedEquipartition::<F, RV32_REGISTER_NUM_LIMBS>::new();
+            for ((addr_space, ptr), metadata) in touched_blocks {
+                if metadata.block_size == AccessMetadata::OCCUPIED {
+                    continue;
+                }
+                assert!(
+                    metadata.block_size >= RV32_REGISTER_NUM_LIMBS as u32,
+                    "How come some block size is less than RV32_REGISTER_NUM_LIMBS, isn't this the alignment?"
+                );
+                let values = (0..metadata.block_size)
+                    .map(|i| self.memory.data.get_f::<F>(addr_space, ptr + i))
+                    .collect::<Vec<_>>();
+                for start in (0..metadata.block_size).step_by(RV32_REGISTER_NUM_LIMBS) {
```

Comment:
I think the step_by size needs to depend on address space? i.e., it is 4 for address space <=3, but 1 for address space 4